### DISCUSSION TITLE
Simple quasar custom component with vue 3 composition API

### DISCUSSION THREAD

**Discussion by bvbellomo** [quasarframework/quasar#12483] [General - Components / Directives / etc] [answered]
Here is the simplest custom component I can write to wrap a q-input and give me v-model binding:

< template>
    <q-input
        outlined
        :model-value="text"
        @update:model-value="(value) => emit('update:text', value)"
        type="text"
    />
</ template>

<script setup lang="ts">
defineProps({
    text: {
        required: false,
        type: String,
    },
});
const emit = defineEmits(['update:text']);
</script>

How do I write something similar with the Vue3 composition API?

Do I use "defineEmits" the same way, import defineEmits from vue?  Or use "$emit"?

Many examples use '$event.target.value' but this gives an error in typescript.  Do I need to find a way to cast this?  Or is this the wrong way to do it?

**Comment by bvbellomo** [comment, marked as answer]
I just found that "script setup" is actually a newer way of using the composition API.  So many different ways to do the same thing with vue, and while we have good examples, they either mix different ways, or at best, don't tell you which one.

**Reply by s-light** [reply]
have you found any guide on *best-practice* for quasar regarding the different styles?

currently i try to refresh my quasar foo and in that learn the composition api way - and i am a little bit puzzled with this...
i found one example in the quasar docs mentioning the *new* `script setup` style:
[SFC with script setup (and Composition API) variant](https://quasar.dev/quasar-plugins/dialog#sfc-with-script-setup-and-composition-api-variant)
not much else...
in the fresh created app it uses something similar to
```
<script>
import { defineComponent, ref } from 'vue'

  export default defineComponent({
    name: 'IndexPage',
    setup () {
        const the_thing = ref("00:00:10");

        return {
            the_thing,
        }
    },
})
</script>
```
after [looking up `defineComponent`](https://vuejs.org/api/general.html#definecomponent) - i think as long as i do not use TS i do not need this extra step.
so maybe i could also just write it this way:
```
<script setup>
import { ref } from 'vue'

const the_thing = ref("00:00:10");
</script>
```
i only do not know if / how i should *define* the `name`..

i hope anybody has any insights into this for me ;-)

sunny greetings
stefan

**Reply by mobsean** [reply]
> i only do not know if / how i should _define_ the `name`..

I´m also transfering my Quasar V1 Project und to V2 and just thought the same.

Why are we using defineComponent() after all? Even when I use TypeScript? I don`t know...

<script setup> is shorter and gives me the same result. 

btw. name is automatically set to the component FileName.
